### PR TITLE
added JNDIObjectLocator and Lazy

### PR DESCRIPTION
**Pull request by lprimak**
Adding JNDI cached / serialized / threaded / immutable JNDI lookup class,
and Lazy, which is a dependency

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -98,6 +98,12 @@
 			<version>2.0.0.Final</version>
 			<scope>provided</scope>
 		</dependency>
+                <dependency>
+                    <groupId>javax.ejb</groupId>
+                    <artifactId>ejb-api</artifactId>
+                    <version>3.0</version>
+                    <scope>provided</scope>
+		</dependency>
```
Comment (BalusC) [inline pom.xml:106]: This is quite a change which changes the current dependency requirements of OmniFaces.

Is this absolutely necessary?

File: pom.xml [hunk 2]
```diff
@@ -119,6 +125,14 @@
 			<scope>provided</scope>
 		</dependency>
 
+                <!-- misc dependencies -->
+                <dependency>
+                    <groupId>org.projectlombok</groupId>
+                    <artifactId>lombok</artifactId>
+                    <version>1.18.12</version>
+                    <scope>provided</scope>
+                </dependency>
```
Comment (BalusC) [inline pom.xml:134]: Please remove this dependency.

File: pom.xml [hunk 3]
```diff
@@ -336,6 +350,7 @@
 					<links>
 						<link>https://javaee.github.io/javaee-spec/javadocs/</link>
 					</links>
+                                        <doclint>none</doclint>
```
Comment (BalusC) [inline pom.xml]: Why this change?

File: src/main/java/org/omnifaces/util/Lazy.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+ * private final Lazy<Object> lazy = new Lazy<>(Object::new);
+ * Object lazyInstance = lazy.get();
+ * }
+ * </pre>
+ *
+ * <a href="https://github.com/flowlogix/flowlogix/blob/master/jakarta-ee/jee-examples/src/main/java/com/flowlogix/examples/LazyExample.java"
+ * target="_blank">Example Code (GitHub)</a>
+ *
+ * @author lprimak
+ * @param <TT> type of object
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/Lazy.java:37]: @Version is missing.

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+ * {@code locator = JNDIObjectLocator.builder().build();
+ * MyEJB myEJB1 = locator.getObject(MyEJB.class);
+ * MyEJB myEJB2 = locator.getObject("java:module/MyEJB");
+ * }
+ * </pre>
+ *
+ * <a href="https://github.com/flowlogix/flowlogix/blob/master/jakarta-ee/jee-examples/src/main/java/com/flowlogix/examples/JndiExample.java"
+ * target="_blank">Example Code (GitHub)</a>
+ *
+ * @author Lenny Primak
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/JNDIObjectLocator.java:61]: @Version is missing.

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    private final boolean noCaching;
+    /**
+     * whether to cache remote EJBs, usually false
+     */
+    private final boolean cacheRemote;
+
+    @Getter(AccessLevel.PACKAGE)
+    private final transient Map<String, Object> jndiObjectCache = new ConcurrentHashMap<>();
+    private final transient Lazy<InitialContext> initialContext = new Lazy<>(this::createInitialContext);
+    private final transient Lock initialContextLock = new ReentrantLock();
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/JNDIObjectLocator.java:103]: I can't make sense of `initialContextLock`, can you explain it?

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+     */
+    private final boolean noCaching;
+    /**
+     * whether to cache remote EJBs, usually false
+     */
+    private final boolean cacheRemote;
+
+    @Getter(AccessLevel.PACKAGE)
+    private final transient Map<String, Object> jndiObjectCache = new ConcurrentHashMap<>();
+    private final transient Lazy<InitialContext> initialContext = new Lazy<>(this::createInitialContext);
```
Comment (BalusC) [inline src/main/java/org/omnifaces/util/JNDIObjectLocator.java:102]: If this were `static` then I can sense the usefulness of `Lazy`, but this is non-static and I can't make sense of its usefulness.

File: pom.xml (hunk 1 above)
Comment (lprimak) [reply pom.xml:106]: This is to import `@Remote` and `@Local` annotations. The only other way I know how to do it is to import the javaee-api dependency.

File: pom.xml (hunk 2 above)
Comment (lprimak) [reply pom.xml:134]: Lombok is not a dependency that 'leaks' into user apps.
I am disappointed that this is only coming up now because I discussed Lombok in my original email.

To me, this is like asking 
"I like OmniFaces, but please remove Java and JSF depencies
because they are \<bloated/slow/dying/ insert your favorite reason here\>"
Lombok is mainly here for the builder pattern, and in turn to make `static final transient` fields work in a serializable class. I know of no other clean method of getting that done.
I also can / will not redesign / rewrite this code from scratch to avoid a "dependency" I just don't have time for that.

Comment (lprimak) [reply pom.xml:134]: see https://stackoverflow.com/questions/2968876/final-transient-fields-and-serialization
the answer with readResolve() and `@Builder` used for copy-constructor

File: pom.xml (hunk 3 above)
Comment (lprimak) [reply pom.xml]: I changed it just now to do all the checks but `-reference` category, because it can't find Lombok-generated builders otherwise and errors out

File: src/main/java/org/omnifaces/util/Lazy.java (hunk 4 above)
Comment (lprimak) [reply src/main/java/org/omnifaces/util/Lazy.java:37]: I am assuming you mean `@since`? Added

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java (hunk 5 above)
Comment (lprimak) [reply src/main/java/org/omnifaces/util/JNDIObjectLocator.java:61]: `@since` added

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java (hunk 6 above)
Comment (lprimak) [reply src/main/java/org/omnifaces/util/JNDIObjectLocator.java:103]: According to this: https://stackoverflow.com/questions/13317848/is-javax-naming-initialcontext-threadsafe
lookups `InitialContext` are not thread-safe and need to be locked.
This is why all lookup calls with this possibly-shared, cached `InitialContext` are protected by this lock so they are not concurrently called.

File: src/main/java/org/omnifaces/util/JNDIObjectLocator.java (hunk 7 above)
Comment (lprimak) [reply src/main/java/org/omnifaces/util/JNDIObjectLocator.java:102]: The initialization is deferred until the field is accessed. Since the initialContext can have different environment variables passed to it from how JNDIObjectLocator was built, it cannot be static.
Also, creating InitialContext is expensive on the app server, as it requires reflection and class resolution at run-time

**Comment by lprimak** [comment]
Thanks for the review!
I am a bit frustrated with all this Lombok pushback. This is not the first time I ran into this,
So I get a bit emotional about this because I feel like people don't really understand it and dismiss it
out-of-hand.
Lombok is really a fantastic tool with very low risk, many contributors, and 10+ years of existence.

**Comment by melloware** [comment]
@lprimak I am a huge Lombok fan myself but I think when it comes to libraries like this one that have been around a long time with lots of code you want to "code in the style of the library".  Meaning you want to code to all look and act the same so introducing Lombok just for this new feature doesn't seem right unless @BalusC were to embrace Lombok for the whole project.  Because then this code doesn't look like any other code in the library.

That being said its not hard to "de-lombok" your classes basically generating the Java code for you for the builder pattern etc...but I am sure you already know that.

Just my two cents...

**Comment by lprimak** [comment]
@melloware thanks for your support!
I do agree with some of your points said above.
However, I don't think that Lombok inherently 'changes' the code style in any significant way necessarily, if
written 'in the spirit' of OmniFaces. I believe that this code fits that description.

Obviously you have to weigh the risks and rewards with any library.
With Lombok, I believe I have done that, and the boilerplate that builder pattern eliminates is well worth
it in this instance.

Also, prior to making this PR, I had an email conversation with @BalusC and mentioned Lombok specifically.
I have heard no pushback so I made the PR.
Obviously it took a lot of work to get to this point as well, unpaid work, so
I am disappointed to hear about this after-the-fact as well, since I wouldn't have done any
of this extra work if I knew Lombok was a no-go ahead of time.

**Comment by melloware** [comment]
No problem.  Thanks for your contribution!

**Comment by BalusC** [comment]
Sorry for the confusion. It was indeed unexpected that you in the end actually changed the setup of an existing library.

I'll look at reimplementing this idea for OmniFaces.

**Comment by BalusC** [comment]
I've retrofitted your code for OmniFaces via https://github.com/omnifaces/omnifaces/commit/f5c4c8c41503ce6cb38f196b01ddeffdb98a1e54

Thank you for your contribution!
